README
XMPlayer Version 0.0.4 - September 15, 2013
================================================================================
Libxenon: d3b3270a3798c1f0b4faeb26e9dfaf24e8d41631

Known Issues
================================================================================
Mounting:
- Endless loading screen: Restart. Not working try formatting using Gui Format32. If this doesn't work place xenon.elf 
on root of internal hdd, and place the "mplayer" folder on the root of a usb (ntfs, fat) then launch
- Sometimes when mounting more than 1 USB, it will not find all of them. Just restart XMPlayer until it does

Playing video:
- Can crash while using play'n'charge kit
- Cannot seek in files over 2 gigabytes (split the files for now)
- Playing .rar can be unstable, sometimes freezing XMPlayer
- Prolonged pausing causes audio to go out of sync, "fix" this by seeking once in the file

Subtitles:
- ASS Subtitle handling can cause freezes and lag. If random crashes in video occurs, set ass=no and change 
vf=ass to #vf=ass in config (ass is disabled by default)
- If your subs doesn't show all the characters, try a different font by replacing subfont.ttf in "mplayer" folder
- If you exit a video with subtitle-text on the screen, the text will still be there if you play a video wihout subtitles, you 
can "fix" this by going into a video with subs and exiting when there is no sutitle-text

Other(s):
- Loading .elf files from the filebrowser on the internal harddrive will crash, this also means using "Restart" will crash if
xenon.elf is only placed on internal harddrive

Not implemented:
- Audio balance, surround sound and optical out are not supported
- Pictures and Music are not supported
- DVD not working (crash)
- No networking!

Install
================================================================================
- Copy all files to the root of an usb stick, launch it by xell or latest dashlaunch

Debug
================================================================================
- Do you get a red screen? Does xmplayer freeze or get stuck at endless loading? and nothing helps?
Read DEBUG so we can fix the problem together

Button Binds
================================================================================
- In File Browser
A: enter
B: up one level
X: sort by
Back: back to homescreen
D-pad: direction

- In Video
A: pause/play
B: exit player
X: show progression
Y: OSD/Menu
Rb: seek 10 min forward
Lb: seek 10 min back
Rt: seek 10 sec forward
Lt: seek 10 sec back
Up: seek 1 min forward
Down: seek 1 min back
Right: seek 10 sek forward
Left: seek 10 sek back

Settings
================================================================================
- Global
Exit action: Shutdown Button bind
Language: Menu language (English/French)

- Audio
Language: Set default audio track of video (i.e mkv with multiple languages)
Volume: Start volume
Soft Volume: Software volume boost

- Video
Frame Dropping: Set default frame dropping option
Vsync: Set default vertical sync option (can cause slow downs if on)

- Subtitles
Color: Color of the subtitle
Border Color: Color of the subtitle border
Code Page: the encoding of the subtitles (latin, baltic, arabic etc.)
Language: Default subtitle language

Thanks to
================================================================================
Ced2911
Aioros
cancerous
GliGli
IceKiller
Juvenal
Natelx
Razkar
sk1080
tuxuser
[c0z]
siz-
Swizzy

Remember to support the xbox 360 legal homebrew community!
www.libxenon.org and #libxenon @ EFNet

HISTORY
Initial:
-Quick fix: Subtitles stays on screen when playing a new video without subs, after exiting a video with subs
-Fixed: "Return to Gui" and starting a video leads to crash 
-Fixed: .avi files not playing after a .mkv, .mov or .mp4 file has been played 
This fix is done via mplayer.conf by adding:
[extension.avi]
demuxer=avi
-Added: "Restart" and "Shutdown" buttons 
-Added: "Subtitle Size" under Settings - Atm you can only change size before starting your first video, after that you need to restart before it takes effect
This is quite simple really, have 3 different fonts sizes from mplayerhq's site. This is temporary until the subscale can be implemented with subfont.ttf
-Vsync is now enabled by default (disable via Gui) <--- CHANGED, see 05/08/2012

24/07/2012:
-Added: A "smart menu" that saves your last selection and path when exiting a menu or playing a video (using y and then exit!)

26/07/2012:
-Fixed: When entering a new folder "Item number/All item number" will not update until moving down on d-pad

29/07/2012:
-Fixed: When paused, the osd buttons (y, xbox, x) unpauses video, only "a" pauses/unpauses now
This is done by adding pausing_keep to the osd binds in mplayer's input.conf
-Changed: The page down icon now only shows, if there are at least more items than a page (10 items pr page) 

30/07/2012:
-Changed: the subtitle quick fix from initial commit has been updated, it looks more stable now

31/07/2012: 
-Added: Resume-playback function: if you stop a video in the middle of playing, it will save last position and resume from there if desired (more than a minute of video has to be played).

05/08/2012:
-Fixed: Osd_show_progression (x) only shows if osd-button (y) hasn't been pressed before
-Fixed: Smart Menu now support multiple devices
This means that the browser will be reset between devices
-Fixed: Lb and Rb doesn't seek -600/+600
Also changed the binds of "r" and "l" to "c" and "v" in input.conf (it didn't seem to detect lowercase l and r)
-Changed: Osd-button now switches from 1 to 3, this means it only needs 1 press to full osd and not 3 presses
-Changed: When pressing the pause-button (a) progression will also show
Bind "a" in input.conf needs to be "pausing_toggle osd_show_progression"
-Removed: Font folder and subtitle size, not needed anymore
Without font-folder the subfont.ttf works fine, with font-scale working
-Added: Support for multiple subtitle languages
Supported language depends on subcp, slang and font. See mplayer\config

06/08/2012:
-Fixed:The D-pad being buggy after a video has been played, every 2nd time you go out and in a menu it won't respond, but the A and B buttons work fine

09/08/2012:
-Added: All libmenu's working functions available in the main osd, so libmenu will no longer be needed

10/08/2012:
-Added: When quiting via libmenu or quit-button (b is binded to quit now), it will now save last position
-Added: Patch from mplayerdev's mailing list which adds support for playing videos inside multivolume uncompressed rar files
Haven't been tested intensively. 

13/08/2012:
-Changed: The B-button doesn't exit to homepage any more, it goes one level up, until root then goes to homepage
-Added: The Back-button goes directly back to homepage
-Added: [extension.rar] demuxer=lavf in mplayer.conf <-- see 17/08/2012

17/08/2012:
-Removed: [extension.rar] demuxer=lavf in mplayer.conf
-Changed: [extension.mp4] demuxer=mov to demuxer=lavf in mplayer.conf, lavf handles .mp4 files better than mov
-Changed: have disabled libmenu, now that osd menu has the same functions, can cause instability
-Added: if a .rar file loads, it now gets the extension of the archived file, so the true extension profile is loaded

27/08/2012:
Added: Sort files. By pressing "x" in the file browser, you can order by Name or Date (day/month/year)

28/08/2012:
Changed: The audio delay option in osd is more logical, a audio_delay of 100 ms doesn't mean audio is 100 ms before video (mplayer standard), but 100 ms after

29/08/2012:
-Changed: Updated libass to 0.10.0
-Changed: ass/ssa is now subtitle default which means outline, color etc. works 
-Added: ass settings can be accessed from the subtitle osd menu (ass can be buggy, video freeze etc.)

04/09/2012:
-Changed: replaced mxml with tinyxml, because of loading problems and it seems lighter

05/09/2012:
-Fixed: some bugs with osd displaying delay as -0 because of it being close to zero
-Added: set ass=yes and vf=ass in mplayer config, with vf=ass it looks stable, haven't experienced freezes. This means it can be disabled and subfont can be used if wished

17/09/2012:
-Added: Global, audio, video and subtitles settings in menu (no need to play with mplayer config files anymore)

01/09/2013:
-Fixed: When in OSD-menu when paused, pressing (Y) won't remove OSD-menu: vo_xenon.c didn't detect the change because flip_page wasn't called, so video-out was not updated until 
video was resumed or seeked in
06/09/2013:
-Fixed: Better support for big USB drives (mount code from XeLL v0.993)

07/09/2013:
- Added: Support for .rmvb files: .rmvb now shows as videofile under Videos, set lavf as it's demuxer in mplayer.conf
- Changed: Disabled balance in osd menu, it doesn't work
- Fixed: Audio- and subdelay shows -0 ms: audio and subtitle delay are not always zero, but because they are floats they can come increadibly close so they can show up as -0.
- Fixed: First time you enter osd menu when paused, it removes it again: Even when osd_level = 3 the OSD layer detected the keypress you made for entering the menu, as 
Y is the OSD button it would remove OSD again. Sometimes OSD would show but then, osd_level = 1 (the player) would detect your next keypress
- Fixed: Volume not changing in audio settings when paused: the OSD wouldn't update the change in volume when video was paused. This was because the OSD used mplayer queue command 

mp_msg.h
/*
 * This file is part of MPlayer.
 *
 * MPlayer is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * MPlayer is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License along
 * with MPlayer; if not, write to the Free Software Foundation, Inc.,
 * 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
 */

#ifndef MPLAYER_MP_MSG_H
#define MPLAYER_MP_MSG_H

#include <stdarg.h>

// defined in mplayer.c and mencoder.c
extern int verbose;

// verbosity elevel:

/* Only messages level MSGL_FATAL-MSGL_STATUS should be translated,
 * messages level MSGL_V and above should not be translated. */

#define MSGL_FATAL 0  // will exit/abort
#define MSGL_ERR 1    // continues
#define MSGL_WARN 2   // only warning
#define MSGL_HINT 3   // short help message
#define MSGL_INFO 4   // -quiet
#define MSGL_STATUS 5 // v=0
#define MSGL_V 6      // v=1
#define MSGL_DBG2 7   // v=2
#define MSGL_DBG3 8   // v=3
#define MSGL_DBG4 9   // v=4
#define MSGL_DBG5 10  // v=5

#define MSGL_FIXME 1  // for conversions from printf where the appropriate MSGL is not known; set equal to ERR for obtrusiveness
#define MSGT_FIXME 0  // for conversions from printf where the appropriate MSGT is not known; set equal to GLOBAL for obtrusiveness

// code/module:

#define MSGT_GLOBAL 0        // common player stuff errors
#define MSGT_CPLAYER 1       // console player (mplayer.c)
#define MSGT_GPLAYER 2       // gui player

#define MSGT_VO 3       // libvo
#define MSGT_AO 4       // libao

#define MSGT_DEMUXER 5    // demuxer.c (general stuff)
#define MSGT_DS 6         // demux stream (add/read packet etc)
#define MSGT_DEMUX 7      // fileformat-specific stuff (demux_*.c)
#define MSGT_HEADER 8     // fileformat-specific header (*header.c)

#define MSGT_AVSYNC 9     // mplayer.c timer stuff
#define MSGT_AUTOQ 10     // mplayer.c auto-quality stuff

#define MSGT_CFGPARSER 11 // cfgparser.c

#define MSGT_DECAUDIO 12  // av decoder
#define MSGT_DECVIDEO 13

#define MSGT_SEEK 14    // seeking code
#define MSGT_WIN32 15   // win32 dll stuff
#define MSGT_OPEN 16    // open.c (stream opening)
#define MSGT_DVD 17     // open.c (DVD init/read/seek)

#define MSGT_PARSEES 18 // parse_es.c (mpeg stream parser)
#define MSGT_LIRC 19    // lirc_mp.c and input lirc driver

#define MSGT_STREAM 20  // stream.c
#define MSGT_CACHE 21   // cache2.c

#define MSGT_MENCODER 22

#define MSGT_XACODEC 23 // XAnim codecs

#define MSGT_TV 24      // TV input subsystem

#define MSGT_OSDEP 25  // OS-dependent parts

#define MSGT_SPUDEC 26 // spudec.c

#define MSGT_PLAYTREE 27    // Playtree handeling (playtree.c, playtreeparser.c)

#define MSGT_INPUT 28

#define MSGT_VFILTER 29

#define MSGT_OSD 30

#define MSGT_NETWORK 31

#define MSGT_CPUDETECT 32

#define MSGT_CODECCFG 33

#define MSGT_SWS 34

#define MSGT_VOBSUB 35
#define MSGT_SUBREADER 36

#define MSGT_AFILTER 37  // Audio filter messages

#define MSGT_NETST 38 // Netstream

#define MSGT_MUXER 39 // muxer layer

#define MSGT_OSD_MENU 40

#define MSGT_IDENTIFY 41  // -identify output

#define MSGT_RADIO 42

#define MSGT_ASS 43 // libass messages

#define MSGT_LOADER 44 // dll loader messages

#define MSGT_STATUSLINE 45 // playback/encoding status line

#define MSGT_TELETEXT 46       // Teletext decoder

#define MSGT_MAX 64


extern char *mp_msg_charset;
extern int mp_msg_color;
extern int mp_msg_module;

extern int mp_msg_levels[MSGT_MAX];
extern int mp_msg_level_all;


void mp_msg_init(void);
int mp_msg_test(int mod, int lev);

#include "config.h"
#if 1
void mp_msg_va(int mod, int lev, const char *format, va_list va);
#ifdef __GNUC__
void mp_msg(int mod, int lev, const char *format, ... ) __attribute__ ((format (printf, 3, 4)));
#   ifdef MP_DEBUG
#      define mp_dbg(mod,lev, args... ) mp_msg(mod, lev, ## args )
#   else
       // only useful for developers, disable but check syntax
#      define mp_dbg(mod,lev, args... ) do { if (0) mp_msg(mod, lev, ## args ); } while (0)
#   endif
#else // not GNU C
void mp_msg(int mod, int lev, const char *format, ... );
#   ifdef MP_DEBUG
#      define mp_dbg(mod,lev, ... ) mp_msg(mod, lev, __VA_ARGS__)
#   else
       // only useful for developers, disable but check syntax
#      define mp_dbg(mod,lev, ... ) do { if (0) mp_msg(mod, lev, __VA_ARGS__); } while (0)
#   endif
#endif /* __GNUC__ */
#else
#define mp_msg_va(...)
#define mp_msg(...)
#define mp_dbg(...)
#endif
const char* filename_recode(const char* filename);

#endif /* MPLAYER_MP_MSG_H */
